### PR TITLE
am9511 - 8080 / 8085 rework

### PR DESCRIPTION
**Pull request by feilipu**
Related to #1944.

### CODE REVIEW

File: libsrc/_DEVELOPMENT/math/float/am9511/c/8085/am9511_asinh.asm [hunk 1]
```diff
@@ -185,41 +185,37 @@
...
+	call	l_f32_add
 	call	_sqrt
-	call	dpush
-	ld	hl,12	;const
+	push	de
+	push	hl
+	ld	hl,8	;const
 	add	hl,sp
-	call	dload
+	call	l_glong
```
Comment (suborb) [inline libsrc/_DEVELOPMENT/math/float/am9511/c/8085/am9511_asinh.asm:211]: Shouldn't this be `call fabs` since `fabs()` is a sccz80 library function?

Comment (feilipu) [reply libsrc/_DEVELOPMENT/math/float/am9511/c/8085/am9511_asinh.asm:211]: TBH, I'm playing with the [makefile](https://github.com/z88dk/z88dk/blob/8827c31cb7847d249f71d274748766590664330f/libsrc/_DEVELOPMENT/math/float/am9511/c/Makefile#L14) to try to get something that looks sensible for use with sccz80/classic/8085. This was the first cut.

At the moment it is compiling against newlib. I know that's incorrect, but I haven't got the right magic to pull in the classic header files and make sure it is using sccz80 to compile. Hints / suggestions on how to get something sensible are welcome. `make` is not my friend.

Comment (suborb) [reply libsrc/_DEVELOPMENT/math/float/am9511/c/8085/am9511_asinh.asm:211]: Ah right, yeah, `zcc +test -clib=8085` should do it for you.

**Comment by feilipu** [comment]
Perplexed by these build errors in the DAI math library. :confused: 
Haven't touched that to my knowledge.

**Comment by suborb** [comment]
> Perplexed by these build errors in the DAI math library. 😕
> Haven't touched that to my knowledge.

My thoughts:

1. daimath uses `math/cimpl/pow_valid.c` to implement `pow()
2. math.h defines pow() to be `pow_callee()`
3. So pow_valid.c ends up exporting` pow_callee()` not `pow()`

**Comment by feilipu** [comment]
> Probably a `#undef pow` needed in `pow_valid.c`

I've matched the classic math headers calling convention for am9511 and math32 to the newlib am9511 and math32 headers.
Does DAI math use either of these two classic math header files?

That's the only place I can guess that could cross over...?

But, otherwise please let me know if I should change `pow_valid.c` (since I don't want to touch something that I don't really understand).

**Comment by suborb** [comment]
Ah, you're right actually - it should be using math_dai32.h, but it's not.

Take a look at the Makefile for daimath32 and you'll see why. It needs to be `--D__MATH_DAI32` not `-D__MATH_AM9511`

**Comment by feilipu** [comment]
Ah. Ok.
Will fix in next push the am.

**Comment by feilipu** [comment]
Chasing down an issue with `%f` format conversion (something to do with the `ftoa.c` helpers).
Pretty sure there's a calling convention mismatch somewhere that is causing a stack problem.

**Comment by feilipu** [comment]
Now an off by 1 in the `%f` format conversion. Or it might be a trashed register.
Could be the last thing.

**Comment by feilipu** [comment]
Fixed the `%d` issue, where the math library `l_long_div_u` doesn't provide the modulus in `a`.
But the `%f` issue remains...

**Comment by feilipu** [comment]
Adding the specific  `l_long_div_m` function, which returns the modulus LSB in `a`, fixes the issue with `%d` and `__print_number`. But it creates an issue where an actual `(unsigned long)x / (unsigned long)y` calculation is required.

There ends up being a double definition of `l_long_div_u`, one from the `l_` source, and one from the `am9511` math source.

Any suggestion on the cleanest way to fix this duplication, while enabling use of the math library option if available?

**Comment by feilipu** [comment]
I'm at a loss now as to why the`%f` format conversion isn't working. Confirmed that `floor()` is working right.
The binary results are right. Just not the print.

Last thing is the calling convention for ftoa (but I fixed that already, I thought :confused: ).
```
floor(2.1666665077) should be 2.0000000000 and we got 2.0000000000
floor(0x3fc00000) should be 0x3f800000 and we got 0x3f800000

floor(-3.8870081901) should be -4.2500000000 and we got -4.2500000000
floor(0xc0670a3d) should be 0xc0800000 and we got 0xc0800000

floor(1000.0110473632) should be 1000.0009765625 and we got 1000.0009765625
floor(0x447a00a5) should be 0x447a0000 and we got 0x447a0000

floor(-1000.0110473632) should be -1001.0009765625 and we got -1001.0009765625
floor(0xc47a00a5) should be 0xc47a4000 and we got 0xc47a4000
```

**Pull request merged**